### ISSUE TITLE
catchError didn't catch the error when using switchMap

### ISSUE THREAD

**Issue by fastrde** [ReactiveX/rxjs#3825]
## Bug Report

**Current Behavior**
when an error is thrown inside a `switchMap` pipe `catchError` in the same pipe didn't catch that error.

**Reproduction**

```js
Rx.of("foo").pipe(
  switchMap(() => Rx.of(undefined).pipe(
    tap(() => console.log("NOT WORKING")),
    mergeMap(boom => boom),
    catchError(err => Rx.throwError("ERROR NOT CATCHED"))
  ))
).subscribe((data) => {
  console.log(data);
},(err) => {
  console.log("ERROR:", err);
})
```
```
Output:
NOT WORKING
ERROR: TypeError: You provided 'undefined' where a stream was expected.
       You can provide an Observable, Promise, Array, or Iterable.
```

**Expected behavior**
the error should be catched and handled by `catchError`

**Environment**
- Runtime: Node v8.11.2
- RxJS version: 6.2.0

**Additional context/Screenshots**
See https://stackoverflow.com/questions/50761948/rxjs6-error-handling-when-using-switchmap
for working vs. not working example.

**Comment by benlesh** [comment]
There isn't an error here. Since you're sending a `throwError` from your `catchError`, it's getting to the error handler in your `subscribe` call, just fine. The error coming out of the `catchError` will also kill the top level chain of observation.

Here's your code in a [Stackblitz](https://stackblitz.com/edit/typescript-fnatwe)

**Comment by fastrde** [comment]
did i missunderstand something? In your stackblitz example the error is 
<img width="556" alt="bildschirmfoto 2018-06-11 um 20 26 51" src="https://user-images.githubusercontent.com/1861723/41249915-b4c73276-6db5-11e8-8d0b-e089d9278eff.png">

When i catch the error and then throw an new one with `ERROR NOT CATCHED` as content in the console should be
```
NOT WORKING
ERROR NOT CATCHED

```
and not the typeerror

**Comment by natmegs** [comment]
It looks like the error is coming from mergeMap, since mergeMap expects to return an Observable value and you're returning just undefined. For example this
```ts
of('foo').pipe(
    switchMap(() => of(undefined).pipe(
        tap(() => console.log('NOT WORKING')),
        mergeMap(boom => of(boom)),
        catchError(err => throwError(err))
    ))
).subscribe(
    data => console.log("SUCCESS: ", data),
    err => console.log("ERROR: ", err),
    () => console.log('complete')
)
```
will return
```ts
NOT WORKING
SUCCESS:  undefined
complete
```

**Comment by fastrde** [comment]
i used something like
```js
.mergeMap(items => items)
.map(item => item.property)
.catch((err) => Observable.throw(new MyError(...properties))
.toArray()
```
to split an array, manipulate the elements and join the array again in rxjs5
and when mergeMap(items => items) throws an error at this point I know that there is no array and the backend delivered something wrong.

In the provided example code I intentionally produce the error to show the potential bug. The behavior I want to discuss in this issue is that the thrown error is not catched by the directly following `catchError`.

I try to explain it better:
```js
Rx.of("foo").pipe( 
  switchMap(() => Rx.of(undefined).pipe( //1 
    tap(() => console.log("NOT WORKING")),
    mergeMap(boom => boom), //2
    catchError(err => {  //3
      console.log("THIS IS NOT LOGGED TO CONSOLE"); 
      return Rx.throwError("ERROR NOT CATCHED");
    })
  ))
).subscribe((data) => {
  console.log(data);
},(err) => {
  console.log("ERROR:", err); //4
})
```

1. The `Rx.of(undefined)` is only for the example in real code this situation happens for example when a backend is not reachable or provide unparsable data
2. at this point the code should split an array into single elements but when the input is `undefined` an error is thrown
3. here the error should be catched and handled (in this example the string "ERROR NOT CATCHED" should be thrown -- **but this didn't work and is the possible bug**
4.` ERROR: ERROR NOT CATCHED` should be logged but `ERROR: TypeError...` is actually logged.

**Comment by cartant** [comment]
@benlesh This looks like a bug - or, at least, inconsistent behaviour - to me.

It's clearer when the working snippet - from the original SO question - is considered.

Further simplified, this snippet catches and handles the error and writes `bar` to the console:

```ts
of("foo").pipe(
    mergeMap(() => undefined),
    catchError(err => of("bar"))
).subscribe(data => console.log(data));
```

However,  the same mechanism - `mergeMap` followed by `catchError` - does not see the error caught when inside `switchMap`:

```ts
of("foo").pipe(
    switchMap(() => of("baz").pipe(
        mergeMap(() => undefined!),
        catchError(err => of("bar"))
    ))
).subscribe(data => console.log(data));
```

In the second snippet, the error is not caught and - in Node - the process dies with this error:

> TypeError: You provided 'undefined' where a stream was expected. You can provide an Observable, Promise, Array, or Iterable.

**Comment by fastrde** [comment]
Looks like the bug is fixed in 6.2.1
